Generate calls to interface methods through resolve helper

MichalStrehovsky opened the pull request:
This changes interface calls when CFG is enabled to use a resolve helper instead of the stub dispatcher. I've also extended the CFG testing a bit to cover more of interface calls in the smoke test.

```csharp
static int Call(IFoo f, int x, int y) => f.Call(x, y);
```

Before this change:

```
00007FF605971D50  push        rbx  
00007FF605971D51  sub         rsp,20h  
00007FF605971D55  mov         qword ptr [rsp+30h],rcx  
00007FF605971D5A  mov         r10d,edx  
00007FF605971D5D  mov         ebx,r8d  
00007FF605971D60  mov         rcx,qword ptr [__InterfaceDispatchCell_repro_Program_IFoo__Call_repro_Program__Call (07FF6059BEBD0h)]  
00007FF605971D67  call        qword ptr [__guard_check_icall_fptr (07FF6059CD558h)]  
00007FF605971D6D  mov         rax,rcx  
00007FF605971D70  mov         rcx,qword ptr [rsp+30h]  
00007FF605971D75  mov         r8d,ebx  
00007FF605971D78  mov         edx,r10d  
00007FF605971D7B  lea         r10,[__InterfaceDispatchCell_repro_Program_IFoo__Call_repro_Program__Call (07FF6059BEBD0h)]  
00007FF605971D82  call        rax  
00007FF605971D84  nop  
00007FF605971D85  add         rsp,20h  
00007FF605971D89  pop         rbx  
00007FF605971D8A  ret  
```

After this change:

```
00007FF704181CF0  sub         rsp,28h  
00007FF704181CF4  lea         r10,[__InterfaceDispatchCell_repro_Program_IFoo__Call_repro_Program__Call (07FF7041CEBD0h)]  
00007FF704181CFB  call        RhpResolveInterfaceMethodFast (07FF703FFF5A0h)  
00007FF704181D00  call        qword ptr [__guard_dispatch_icall_fptr (07FF7041DD560h)]  
00007FF704181D06  nop  
00007FF704181D07  add         rsp,28h  
00007FF704181D0B  ret  
```

MichalStrehovsky committed:
Generate calls to interface methods through resolve helper

jakobbotsch committed:
Expand resolver during CFG expansion

jakobbotsch committed:
Revert register constraint changes

jakobbotsch committed:
Nit

MichalStrehovsky commented:
/azp run runtime-nativeaot-outerloop

MichalStrehovsky committed:
Revert "Revert register constraint changes"

This reverts commit 7f879d6accfbc2ec5b486b811b16e4e65e3164ee.

MichalStrehovsky committed:
Scary hack

MichalStrehovsky commented:
For this not to be a throughput regression, we need to figure out a clean way to implement the hack in last two commits.

I ran JSON serialization with/without this PR. Having to preserve argument registers before calling the resolve helper costs us about 1% in throughput.

PR-hack is the hack where we allow RyuJIT to assume argument registers are not clobbered. We better not take a GC in the slow path.
PR is this PR without that hack.
Baseline is the vanilla CFG build.

Style | Attemp1 ms | Attemp2 ms | Attemp3 ms | Attemp4 ms | Attemp5 ms
-- | -- | -- | -- | -- | --
PR-hack | 1103 | 1105 | 1096 | 1102 | 1103
PR | 1125 | 1126 | 1139 | 1127 | 1134
Baseline | 1112 | 1105 | 1111 | 1104 | 1106

Do we have a way to do this hack cleanly that doesn't involve building a new universal transition?

MichalStrehovsky committed:
Universal transition

MichalStrehovsky committed:
Arm64 and fixes

MichalStrehovsky committed:
Testing

MichalStrehovsky commented:
/azp run runtime-nativeaot-outerloop

MichalStrehovsky committed:
Merge branch 'main' into resolvehelper

MichalStrehovsky committed:
Update jiteeversionguid.h

MichalStrehovsky commented:
@dotnet/ilc-contrib this is ready for review

@jakobbotsch who would be the best to review the JIT side?

jakobbotsch commented on src/coreclr/inc/jithelpers.h:
```diff
@@ -309,6 +309,7 @@
     JITHELPER(CORINFO_HELP_JIT_REVERSE_PINVOKE_EXIT_TRACK_TRANSITIONS,  JIT_ReversePInvokeExitTrackTransitions,  METHOD__NIL)
 
     JITHELPER(CORINFO_HELP_GVMLOOKUP_FOR_SLOT, NULL, METHOD__NIL)
+    JITHELPER(CORINFO_HELP_INTERFACELOOKUP_FOR_SLOT, NULL, METHOD__NIL)
```
How hard would it be to add an implementation for coreclr? Currently [jit-cfg](https://dev.azure.com/dnceng-public/public/_build?definitionId=155) is likely to be broken. It would be nice to make it work without having to make the JIT side codegen different from NAOT pattern.

Alternatively I guess it would be even more ideal if we switched the testing to a CFG-enabled outerloop NAOT run now that the support is more mature... But that's probably a bit more work.

jakobbotsch replied:
I am not totally sure what the difference between CID and VSD is, but from the JIT codegen standpoint it looks identical to me.

In the worst case I would be ok with guarding the transformation in lowering under a NAOT check.

MichalStrehovsky replied:
@jkotas if I wanted to implement a CORINFO_HELP_INTERFACELOOKUP_FOR_SLOT on CoreCLR, would I need to somehow call `VirtualCallStubManager::ResolveWorker`? It's not exactly clear to me where I would find the parameters to call it so wondering if there's a better way.

jkotas replied:
I have pushed a commit into this PR that shows what it may look like on win x64. It calls the slow resolve helper every time. If it is too slow for the test, it is possible to improve the perf by adding a NAOT-like fast path once David's PR https://github.com/dotnet/runtime/pull/111771/ goes in.

MichalStrehovsky replied:
> I have pushed a commit into this PR that shows what it may look like on win x64

Thank you ❤️! I'll kick off a JIT-cfg run to see if the perf is acceptable. I think we'd basically only care if this more than quadruples the time it takes for the CI to run. These legs run very infrequently.

MichalStrehovsky commented:
/azp list

MichalStrehovsky commented:
/azp run jit-cfg

jkotas commented:
Runtime changes LGTM

jkotas committed:
Add JIT_InterfaceLookupForSlot helper for Win x64

MichalStrehovsky commented:
/azp run jit-cfg

jkotas commented:
Number of tests are failing with: Assertion failed 'false && "found use of a node that is not in the LIR sequence"' in

MichalStrehovsky commented:
> Number of tests are failing with: Assertion failed 'false && "found use of a node that is not in the LIR sequence"' in

Oh, which legs had that? I've been looking at the jit-cfg Win-x64 leg failing with:

```
Assert failure(PID 8692 [0x000021f4], Thread: 1020 [0x03fc]): !CREATE_CHECK_STRING(!"Detected use of a corrupted OBJECTREF. Possible GC hole.")

CORECLR! `Object::ValidateInner'::`1'::catch$13 + 0x10B (0x00007fff`e813024b)
CORECLR! CallSettingFrame_LookupContinuationIndex + 0x20 (0x00007fff`e8075b50)
CORECLR! _FrameHandler4::CxxCallCatchBlock + 0x1DE (0x00007fff`e80739fe)
NTDLL! RtlCaptureContext + 0x3C3 (0x00007fff`f37aa173)
CORECLR! Object::ValidateInner + 0xFE (0x00007fff`e79771de)
CORECLR! Object::Validate + 0xA1 (0x00007fff`e79770a1)
CORECLR! OBJECTREF::OBJECTREF + 0x9E (0x00007fff`e796f0ee)
CORECLR! PreStubWorker + 0x26D (0x00007fff`e79ab27d)
CORECLR! ThePreStub + 0x55 (0x00007fff`e7e69395)
<no module>! <no symbol> + 0x0 (0x00007fff`88623920)
    File: D:\a\_work\1\s\src\coreclr\vm\object.cpp:638
    Image: C:\h\w\A1C70906\p\corerun.exe
```

But maybe the JIT assert is related to that? (I wasn't able to find where the GC hole could be yet.)

jkotas commented:
> Oh, which legs had that?

For example: https://dev.azure.com/dnceng-public/public/_build/results?buildId=957514&view=ms.vss-test-web.build-test-results-tab&runId=25483200&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab&resultId=119886

> I wasn't able to find where the GC hole could be yet.

My top suspect would be the GC info around the callsite of interface lookup helper. I would check the generated GC info in the jit dump. (The GCInfo command in windbg is broken at the moment.)

jakobbotsch commented:
I can take a look at the JIT assert.

MichalStrehovsky commented:
> I can take a look at the JIT assert.

Did you have a chance to look at it?

jakobbotsch committed:
Clear out call cookie when changing VSD call to indirect

jakobbotsch committed:
Nit

jakobbotsch committed:
Probable fix for GC issues

jakobbotsch committed:
Undo accidental change

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into pr-112406

jakobbotsch committed:
Update JIT-EE GUID after merge

jakobbotsch committed:
Run jit-format

jakobbotsch commented:
I pushed a commit to fix the JIT assert and also a probable fix for the GC stress issue (only based on review, haven't tested that one). Will kick off the CFG job again.

jakobbotsch commented:
/azp run jit-cfg

jakobbotsch commented:
There's still gcstress failures. I'm guessing there are some more fundamental issues around trying to give this helper a non standard calling convention where GC refs could be in the argument registers. I suspect they are not properly being reported and reflected back to the JIT on return.

jkotas commented:
> I suspect they are not properly being reported

The GC info produced by the JIT does not look right.

GC info without CFG - notice that tracking of rcx/rdx/r8/r9 is stopped after the interface call:
```
IN0003: 00000D mov      rcx, 0x1F972400BE8      ; data for Program:f
IN0004: 000017 mov      rcx, gword ptr [rcx]
                            ; gcrRegs +[rcx]
IN0005: 00001A mov      r9, 0x1F972400BF0      ; data for Program:o
IN0006: 000024 mov      r9, gword ptr [r9]
                            ; gcrRegs +[r9]
IN0007: 000027 mov      rdx, r9
                            ; gcrRegs +[rdx]
IN0008: 00002A mov      r8, r9
                            ; gcrRegs +[r8]
IN0009: 00002D mov      r11, 0x7FF91B770070      ; code for IFace:M(System.Object,System.Object,System.Object):this
IN000a: 000037 call     [r11]IFace:M(System.Object,System.Object,System.Object):this
                            ; gcrRegs -[rcx rdx r8-r9]
                            ; gcr arg pop 0
```

GC info with CFG - notice that tracking of rcx/rdx/r8/r9 is stopped before the interface call:
```
IN0003: 00000D mov      r11, 0x1F65D000BE8      ; data for Program:f
IN0004: 000017 mov      rcx, gword ptr [r11]
                            ; gcrRegs +[rcx]
IN0005: 00001A mov      r11, 0x1F65D000BF0      ; data for Program:o
IN0006: 000024 mov      r9, gword ptr [r11]
                            ; gcrRegs +[r9]
IN0007: 000027 mov      rdx, r9
                            ; gcrRegs +[rdx]
IN0008: 00002A mov      r8, r9
                            ; gcrRegs +[r8]
IN0009: 00002D mov      r11, 0x7FF8FB8B0070      ; code for IFace:M(System.Object,System.Object,System.Object):this
recordRelocation: 00007FF8FBBB3CE8 (rw: 000001F600575220) => 00007FF95B3DAB40, type 16 (IMAGE_REL_BASED_REL32), delta 0
IN000a: 000037 call     CORINFO_HELP_INTERFACELOOKUP_FOR_SLOT
                            ; gcrRegs -[rcx rdx r8-r9] <--- this looks wrong, it should be after call [CORINFO_HELP_DISPATCH_INDIRECT_CALL]
                            ; gcr arg pop 0
recordRelocation: 00007FF8FBBB3CEE (rw: 000001F600575226) => 00007FF95B98DCF0, type 16 (IMAGE_REL_BASED_DISP32), delta 0
IN000b: 00003C call     [CORINFO_HELP_DISPATCH_INDIRECT_CALL]
                            ; gcr arg pop 0
```

Test program source
```

class Program
{
    static IFace f = new C();
    static object o = new object();

    static void Main() 
    {
        f.M(o,o,o);
    }
}

interface IFace
{
   void M(object a, object b, object c);
}

class C : IFace
{
    public void M(object a, object b, object c) { }
}
```

jakobbotsch commented:
I can look into that, but do you expect that this is going to make things work? My general wondering is if the runtime would know to report and relocate the GC refs in the argument registers of the transition frame if GC happens while inside the new helper.

jakobbotsch committed:
Fix Jan's bug, fix tailcall test

jakobbotsch committed:
Fix build on targets without custom calling convention for helper

jakobbotsch commented:
/azp run jit-cfg

jakobbotsch committed:
JIT: Simplify internal GC tracking structures

Track all integer registers for calls in `regPtrDsc`. This does not cost
any extra memory and it saves us from going back and forth between an
intermediate format. It also unblocks proper GC reporting for helper
calls that are GC reported with non-standard calling convention.

jkotas committed:
Delete unnecessary TAILJMP_RAX

jkotas commented:
> My general wondering is if the runtime would know to report and relocate the GC refs in the argument registers of the transition frame if GC happens while inside the new helper.

The runtime knows how to do it for the existing VSD helper that tailcalls, so it should be able to do it for helper that returns as well. The new helper uses the same GC reporting setup as the existing VSD helper that tailcalls.

In order for this to work, the JIT has to setup all arguments before calling the new helper. It does not seem to be happening in all cases. For example, I see the following codegen for a call from `Dictionary.FindValue` to `System.Collections.Generic.IEqualityComparer<string>.Equals(string,string)`:

```
...
00007ff8`f8b2d475 4c8bd8          mov     r11,rax
00007ff8`f8b2d478 498b1424        mov     rdx,qword ptr [r12]
00007ff8`f8b2d47c 488bcf          mov     rcx,rdi
00007ff8`f8b2d47f e8bcd6825f      call    coreclr!JIT_InterfaceLookupForSlot (00007ff9`5835ab40)
00007ff8`f8b2d484 488bcf          mov     rcx,rdi <--- unnecessary duplication, but it does not hurt
00007ff8`f8b2d487 4c8bc6          mov     r8,rsi <--- this needs to be before JIT_InterfaceLookupForSlot 
00007ff8`f8b2d48a ff156008de5f    call    qword ptr [00007ff9`5890dcf0] // JIT_DispatchIndirectCall
00007ff8`f8b2d490 85c0            test    eax,eax
...
```

jakobbotsch commented:
> In order for this to work, the JIT has to setup all arguments before calling the new helper. It does not seem to be happening in all cases. For example, I see the following codegen for a call from `Dictionary.FindValue` to `System.Collections.Generic.IEqualityComparer<string>.Equals(string,string)`:

That's expected with how this is implemented in the JIT. The JIT just considers this helper to not trash the argument registers. It is free to leave any GC ref in those registers across the call. It is not considering the "resolve + dispatch" sequence as one unit that takes the arguments of the final call. (I am not sure if this would be feasible on arm64 that uses "resolve + validate + call" sequence).
What the JIT needs from the VM side here is that the VM side uses its reported GC information to update the argument registers of the transition frame (like it would already be doing for callee saves). Is it possible to make it do that?

jkotas committed:
Depend on the GCInfo produced by the JIT for reporting of arg registers

jkotas commented:
> What the JIT needs from the VM side here is that the VM side uses its reported GC information to update the argument registers of the transition frame (like it would already be doing for callee saves). Is it possible to make it do that?

Yes, I have pushed a commit to do that (some of the changes in that commit are quick hack that needs cleaning up if it works).

jakobbotsch committed:
Remove check

jakobbotsch commented:
/azp run jit-cfg

risc-vv commented:
@dotnet/samsung Could you please take a look? These changes may be related to riscv64.

risc-vv commented:
@dotnet/samsung Could you please take a look? These changes may be related to riscv64.

risc-vv commented:
<details><summary>42343da2 is being scheduled for building and testing</summary>

GIT: `42343da2b60dee7fe1cb10fc0fcd9ee4ec9fcfb1`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

risc-vv commented:
@dotnet/samsung Could you please take a look? These changes may be related to riscv64.

risc-vv commented:
<details><summary>42343da2 is being scheduled for building and testing</summary>

GIT: `42343da2b60dee7fe1cb10fc0fcd9ee4ec9fcfb1`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

risc-vv commented:
<details><summary>42343da2 is being scheduled for building and testing</summary>

GIT: `42343da2b60dee7fe1cb10fc0fcd9ee4ec9fcfb1`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

risc-vv commented:
<details><summary>42343da2 is being scheduled for building and testing</summary>

GIT: `42343da2b60dee7fe1cb10fc0fcd9ee4ec9fcfb1`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

MichalStrehovsky commented:
> Really? That would surprise me. JIT should make sure this goes through CFG. That should definitely be fixed.
> What's the repro for this?

Ah, yeah, this goes through a CFG check. But your suggestion is to go through two CFG checks for a single interface call?

jakobbotsch commented:
> Ah, yeah, this goes through a CFG check. But your suggestion is to go through two CFG checks for a single interface call?

Yes, it would end up going through two CFG checks. I don't know how bad the impact on performance would be.

jkotas commented:
> it seems like a lot of risk to me.

This change looks risky to me for .NET 10 at this point regardless of the implementation strategy.

risc-vv commented:
@dotnet/samsung Could you please take a look? These changes may be related to riscv64.
